pkg/util/bulk/aggregator_stats.go

```diff
@@ -0,0 +1,152 @@
+// licenses/APL.txt.
+
+package bulk
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
+)
+
+var tagToAggregatorEventConstructor map[string]func(aggEvent []byte) (TracingAggregatorEvent, error)
+
+// RegisterAggregatorEventConstructor registers a constructor method for a
+// particular TracingAggregatorEvent identified by its tag.
+func RegisterAggregatorEventConstructor(
+	aggEventTag string, constructor func(content []byte) (TracingAggregatorEvent, error),
+) {
+	if tagToAggregatorEventConstructor == nil {
+		tagToAggregatorEventConstructor = make(map[string]func(aggEvent []byte) (TracingAggregatorEvent, error))
+	}
+	_, ok := tagToAggregatorEventConstructor[aggEventTag]
+	if ok {
+		panic(fmt.Sprintf("duplicate stringers registered for %s", aggEventTag))
+	}
+	tagToAggregatorEventConstructor[aggEventTag] = constructor
+}
+
+// flushTracingStats persists the following files to the `system.job_info` table
+// for consumption by job observability tools:
+//
+// - A file per node, for each aggregated TracingAggregatorEvent. These files
+// contain the machine-readable proto bytes of the TracingAggregatorEvent.
+//
+// - A text file that contains a cluster-wide and per-node summary of each
+// TracingAggregatorEvent in its human-readable format.
+func flushTracingStats(
+	ctx context.Context,
+	jobID jobspb.JobID,
+	db isql.DB,
+	perNodeStats map[execinfrapb.ComponentID]map[string][]byte,
+) error {
+	clusterWideAggregatorStats := make(map[string]TracingAggregatorEvent)
+	asOf := timeutil.Now().Format("20060102_150405.00")
+
+	var clusterWideSummary bytes.Buffer
+	for component, tagToEvent := range perNodeStats {
+		clusterWideSummary.WriteString(fmt.Sprintf("## SQL Instance ID: %s; Flow ID: %s\n\n",
+			component.SQLInstanceID.String(), component.FlowID.String()))
+		for tag, event := range tagToEvent {
+			// Write a proto file per tag. This machine-readable file can be consumed
+			// by other places we want to display this information egs: annotated
+			// DistSQL diagrams, DBConsole etc.
+			filename := fmt.Sprintf("%s/%s/%s.binpb",
+				component.SQLInstanceID.String(), tag, asOf)
+			if err := jobs.WriteExecutionDetailFile(ctx, filename, event, db, jobID); err != nil {
+				return err
+			}
+
+			// Construct a single text file that contains information on a per-node
+			// basis as well as a cluster-wide aggregate.
+			clusterWideSummary.WriteString(fmt.Sprintf("# %s\n", tag))
+
+			f, ok := tagToAggregatorEventConstructor[tag]
+			if !ok {
+				return errors.Newf("could not find constructor for AggregatorEvent %s", tag)
+			}
+			aggEvent, err := f(event)
+			if err != nil {
+				return err
+			}
```

Comment:
Should this be some sort of best-effort thing? Perhaps write the error to the file rather than the event and keep going?

Reply:
It should, changed to be best-effort unless we fail to write to system.job_info in which case the method stops trying to write anymore files.